codec/src/main/java/io/netty/handler/codec/compression/ZstdEncoder.java

```diff
@@ -0,0 +1,327 @@
+        }
+
+        if (preferDirect) {
+            return ctx.alloc().ioBuffer(targetBufSize, targetBufSize);
+        } else {
+            return ctx.alloc().heapBuffer(targetBufSize, targetBufSize);
+        }
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws Exception {
+        if (finished) {
+            if (!out.isWritable(in.readableBytes())) {
+                // out should be EMPTY_BUFFER because we should have allocated enough space above in allocateByteBuffer.
+                throw ENCODE_FINISHED_EXCEPTION;
+            }
+            out.writeBytes(in);
+            return;
+        }
+
+        final ByteBuf buffer = this.buffer;
+        int length;
+        while ((length = in.readableBytes()) > 0) {
+            final int nextChunkSize = Math.min(length, buffer.writableBytes());
+            in.readBytes(buffer, nextChunkSize);
+
+            if (nextChunkSize <= MIN_BLOCK_SIZE) {
+                writeUnCompressedData(out);
+                return;
+            }
+
+            if (!buffer.isWritable()) {
+                flushBufferedData(out);
+            }
+        }
+        if (buffer.isReadable()) {
+            flushBufferedData(out);
+        }
+    }
+
+    private void flushBufferedData(ByteBuf out) {
+        final int flushableBytes = buffer.readableBytes();
+        if (flushableBytes == 0) {
+            return;
+        }
+        checksum.reset();
+        checksum.update(buffer, buffer.readerIndex(), flushableBytes);
+        final int check = (int) checksum.getValue();
+
+        final int bufSize = (int) Zstd.compressBound(flushableBytes) + HEADER_LENGTH;
+        out.ensureWritable(bufSize);
+        final int idx = out.writerIndex();
+        int compressedLength;
+        try {
+            ByteBuffer outNioBuffer = out.internalNioBuffer(idx + HEADER_LENGTH, out.writableBytes() - HEADER_LENGTH);
+            compressedLength = Zstd.compress(
+                    outNioBuffer,
+                    buffer.internalNioBuffer(buffer.readerIndex(), flushableBytes),
+                    DEFAULT_COMPRESS_LEVEL);
+        } catch (Exception e) {
+            throw new CompressionException(e);
+        }
+        final int blockType;
+        if (compressedLength >= flushableBytes) {
+            blockType = BLOCK_TYPE_NON_COMPRESSED;
+            compressedLength = flushableBytes;
+            out.setBytes(idx + HEADER_LENGTH, buffer, 0, flushableBytes);
+        } else {
+            blockType = BLOCK_TYPE_COMPRESSED;
+        }
+
+        out.setInt(idx, ZSTD_NUMBER);
+        out.setByte(idx + TOKEN_OFFSET, (byte) (blockType | compressionLevel));
+        out.setIntLE(idx + COMPRESSED_LENGTH_OFFSET, compressedLength);
+        out.setIntLE(idx + DECOMPRESSED_LENGTH_OFFSET, flushableBytes);
+        out.setIntLE(idx + DEFAULT_CHECKSUM_OFFSET, check);
+        out.writerIndex(idx + HEADER_LENGTH + compressedLength);
```

Comment:
> This is a specific custom protocol, not an implementation that follows a certain framework

Thanks for confirming. I notice that this PR follows the format similar to `Lz4FrameEncoder`, but no other library follows the same format for Zstd.

The reason why we use this format for LZ4 codec is because `lz4-java` [1] was the only one java implementation of LZ4 back in 2014. Therefore, we decided to follow the same format as `LZ4BlockOutputStream` [2] to make netty implementation compatible with any other java system that may use the same library. We verify this compatibility in `Lz4FrameEncoderTest` and `Lz4FrameDecoderTest`.

Time flies and I see that the official LZ4 website [3] now recommends Apache Commons Compress as an interoperable java LZ4 port. Apache added support for LZ4 in 2017 [4].

Zstd has a broader scope and applicability compare to LZ4. It's an official standard [5] that defines the format of compressed frames and the frame header. The standard also defines `application/zstd` HTTP media-type and `zstd` content-encoding. The Zstd algorithm is a lot more used in the world for different use-cases, including HTTP.

Taking that into account, I would like to discuss with you the motivation for adding Zstd codec with a custom header that is not supported by any other non-netty system. Could you please describe your use-case?

It will be highly appreciated if we can follow the standard and implement a codec with the interoperable Zstd format. It will allow us to reuse your work in the HTTP codec and communicate with any other system that uses Zstd. And we will be able to use `zstd-jni` [6] `Zstd[Input|Output]Stream` to verify the compatibility.

[1] https://github.com/lz4/lz4-java
[2] https://github.com/lz4/lz4-java/blob/master/src/java/net/jpountz/lz4/LZ4BlockOutputStream.java
[3] https://lz4.github.io/lz4/
[4] https://github.com/apache/commons-compress/commits/master/src/main/java/org/apache/commons/compress/compressors/lz4/FramedLZ4CompressorInputStream.java
[5] https://tools.ietf.org/html/rfc8478
[6] https://github.com/luben/zstd-jni

Reply:
Thanks for your suggestion, so do you mean that the headers needs to be removed, or other changes are needed?

Reply:
I'm suggesting to implement a header format defined by the Zstd RFC [1] instead of defining a custom header that can not be understood by any other non-netty system. We should have a strong reason to implement a custom header instead of following the RFC. If you have this reason, please provide more context about why you chose the current header format.

`zstd-jni` will be very helpful to do most of the work. In netty, look at how `ZlibEncoder` and `ZlibDecoder` are implemented. The gzip/deflate have a similar format to Zstd.

[1] https://tools.ietf.org/html/rfc8478

Reply:
The main point here that whatever `ZstdEncoder` produce, other systems should be able to decompress and vice versa. For the test we can use `ZstdInputStream` and `ZstdOutputStream` classes from zstd-jni. See `Bzip2EncoderTest` and `Bzip2DecoderTest` as an examples of how to verify output of the encoder/decoder with Input/Output streams.

Reply:
Before you address other comments, let's finish this discussion and come to some agreement. Because if we decide to follow the standard it will require significant changes of the current encoder and decoder implementation and my other comments won't be relevant anymore.

// cc @normanmaurer @trustin wdyt about having a non-standard header for Zstd?

Reply:
@skyguard1 as @idelpivnitskiy mentioned we need to follow the RFC and implement the headers the way its defined there. Otherwise it will be impossible to interopt with other implementations out in the wild